macros/support/src/class/iunknown_impl.rs

```diff
@@ -35,25 +35,28 @@ impl IUnknownAbi {
         quote! {
             unsafe extern "stdcall" fn release(this: #this_ptr) -> u32 {
                 #munge
-                #class_name::release(&this)
+                let count = #class_name::release(&munged);
+                if count == 1 {
```

Comment:
I'm not sure how I feel about this. We're relying on a ref count of 1 meaning that the only handle to the class is the original one which should have been "forgotten" through the use of `ManuallyDrop`. If the user forgets to wrap the class in a `ManuallyDrop`, and calls `query` to get an interface then it will have a ref count of 1 (the handle to the interface requested in `query`. When `release` is called on that interface the count will be 0. This is unfortunate as it leaks memory, though it is not unsafe. 

An alternate implementation would drop the class when the count is 0. However, normally, this would leak memory in the normal case of "forgetting" the class when created by a class factory.

Reply:
It's actually also possible to have a dangling pointers if this function is called with a class that has not been "forgotten":

If I return an interface from IClassFactory::create_instance that has been created from a class struct that was then dropped, it's ref count will be 1 (just the handle to interface instead of 2 for the handle and the original class object). If I then make another handle, and it gets released through the call to release ABI function, it will free the memory even though there is still a handle to an interface out there. 

This requires improperly implementing an unsafe function though so I don't think it's not an obviously wrong way to go.

Reply:
The usual way to handle this is for the "make" function to allocate the object and return the only reference to the object as a COM interface. That COM interface holds that first reference count.

Reply:
Yes, this is what we're currently doing, though the user can allocate on their own (i.e., there's nothing stopping them other than us documenting that they shouldn't). Our "make" function is called `alloc` and it does just this. I'm looking for a way to prevent users from holding on to COM classes at all since they should use `alloc`.

Reply:
Ah, so in C++/WinRT we have a virtual function called [use_make_function_to_create_this_object](https://kennykerr.ca/2019/05/08/cppwinrt-diagnosing-direct-allocations/) that solves this problem. Perhaps in Rust we could have the extra fields needed for COM (e.g. vfptrs and ref count) such that they are not `Default` so you cannot simply construct the object yourself.

Reply:
Yea I'm currently thinking how we want to do this. The code is being generated by the macro so it's currently as if the user wrote the code themselves (i.e., everything the macro produces, the user must also be able to write themselves). I'm thinking about how we can give the macro the ability to create the class but not give this ability to the user.

Reply:
I pushed a change that should help. 

It's still technically possible inside of the module where the class is declared to instantiate the class on the stack, but this has be done by hand and is not trivial (it's about 225 lines of total code in the basic server example). The class is not Default so there's no way to instantiate that way (though you can call `MyClass::allocate_default()` to get a proper heap allocated version where all the user fields are defaulted). Outside of the current module it's not possible at all to stack allocate the class. 

Obviously we need docs to clarify how this all should work, but it's now highly unlikely that the user will do the wrong thing.